include/envoy/http/filter.h

```diff
@@ -308,6 +308,11 @@ class StreamDecoderFilterCallbacks : public virtual StreamFilterCallbacks {
    * @return the buffer limit the filter should apply.
    */
   virtual uint32_t decoderBufferLimit() PURE;
+
+  /**
+   * This routine disarms the request timer owned by the connection manager, if it exists.
+   */
+  virtual void upstreamRequestComplete() PURE;
```

Comment:
Ok, having read through this PR, I'd like to hear a bit of a justification for adding this method.  Adding something to streamDecoderFilterCallbacks is making more complicated a core interface of envoy, so I'd like there to be a justification for adding it in terms of envoy production functionality.  It *looks* like this is just a method to make testing easier.  Is that right?  If so, let's try and find a different way to do the same thing that doesn't involve making this interface more complex.

Reply:
+1, I don't think this belongs in the interface. Can we figure out how to remove?

IMO the way to remove this is to say the request timer gets disarmed when the HCM gets the entire request and not involve other filters. If other filters decide to pause or do other things it's up to them?

Reply:
This is pretty much what I had before, as per [this thread](https://github.com/envoyproxy/envoy/pull/4456#discussion_r219581980). This doesn't make testing easier, but rather avoids a spammy check to `stream_info` (formerly `request_info`) on every bit of encoding.

@zuercher can you weigh in?

Reply:
I'm inclined to go for the latter, because then having a request idle timeout and a response idle timeout covers full stream lifetime, where if do the former, I think there would be a gap between request and response idle timeout where any filter doing async operation would leave a potential gap not covered by idle timers where connections could leak.

I think given we also have the keep-alive idle timeout timer we could go either way, but I'd mildly prefer the request/response timers give full coverage. SG?

Reply:
I'm fine either way as long as it's well documented. (We should make sure this is clearly spelled out in the docs for the feature.)